File: api/arretes.py
"""
PopPilot API — arrêtés ENREGISTRÉS par domaine (règle du calendrier).

  GET /arretes/{domaine}      dates d'arrêté chargées, la plus récente d'abord

Règle d'affichage (CDG, 25/09/2026) : l'utilisateur choisit N'IMPORTE QUELLE date dans le
calendrier ; l'écran montre le DERNIER arrêté enregistré à cette date (arrêté ≤ date choisie),
et affiche cette date-là. Si la date précède tout import, le plus ancien arrêté enregistré.
Ex. : 5 juillet → chiffres ET date du 30 juin ; 30 juin → 30 juin.

La résolution se fait dans l'écran (web/src/lib/arretes.ts) à partir de cette liste ; ce module
ne fait que dire ce que la base contient vraiment (date_arrete fait foi, jamais date_snapshot).
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select

from socle.schema import (CompteResultatAgence, FaitBalance, FaitCredit, FaitEpargne,
                          get_session)

from auth_supabase import utilisateur_courant

routeur = APIRouter(tags=["arretes"])

# domaine → table dont les arrêtés gouvernent l'écran
DOMAINES = {
    "credit": FaitCredit,                       # crédit, productivité, primes
    "balance": FaitBalance,                     # comptabilité, indicateurs, budget
    "epargne": FaitEpargne,
    "compte_resultat_agence": CompteResultatAgence,
}


def arretes_enregistres(domaine: str, db_path=None) -> list:
    table = DOMAINES[domaine]
    s = get_session(db_path) if db_path else get_session()   # défaut = base de l'API
    try:
        return s.execute(select(table.date_arrete).distinct()
                         .order_by(table.date_arrete.desc())).scalars().all()
    finally:
        s.close()


@routeur.get("/arretes/{domaine}")
def endpoint_arretes(domaine: str, user: dict = Depends(utilisateur_courant)):
    if domaine not in DOMAINES:
        raise HTTPException(422, f"domaine inconnu : {domaine} (attendu {', '.join(DOMAINES)})")
    # Des dates seulement, aucun montant : visibles par tout profil connecté.
    return {"domaine": domaine, "arretes": [d.isoformat() for d in arretes_enregistres(domaine)]}

File: web/src/lib/arretes.ts
/**
 * PopPilot — regle du calendrier (CDG, 25/09/2026).
 *
 * L'utilisateur choisit N'IMPORTE QUELLE date. L'ecran montre le DERNIER arrete
 * enregistre a cette date (arrete <= date choisie) et AFFICHE cette date-la :
 * 5 juillet → chiffres et date du 30 juin ; 30 juin → 30 juin. Si la date precede
 * tout import : le plus ancien arrete enregistre. Sans date : le plus recent.
 *
 * Les arretes viennent de GET /arretes/{domaine} (ce que la base contient vraiment).
 */
import { appelerApi } from "@/lib/api";
import { dateArreteValide } from "@/lib/format";

export type DomaineArretes = "credit" | "balance" | "epargne" | "compte_resultat_agence";

export type ArreteResolu = {
  /** Arrete effectivement affiche (null : rien d'enregistre). */
  arrete: string | null;
  /** Date saisie dans le calendrier, si elle differe de l'arrete affiche. */
  demande: string | null;
  /** Arretes enregistres, le plus recent d'abord. */
  disponibles: string[];
  /** La liste n'a pas pu etre lue (API muette) : l'ecran garde la date saisie. */
  erreur: string | null;
};

/** Dernier arrete <= date choisie ; sinon le plus ancien ; sans date, le plus recent. */
export function resoudreArrete(demande: string | null | undefined, disponibles: string[]): string | null {
  if (disponibles.length === 0) return null;
  const tries = [...disponibles].sort().reverse();
  if (!demande || !dateArreteValide(demande)) return tries[0];
  return tries.find((a) => a <= demande) ?? tries[tries.length - 1];
}

/** Lit les arretes du domaine et resout la date saisie. */
export async function arreteAffiche(
  domaine: DomaineArretes,
  demande: string | null | undefined,
  jeton: string | null | undefined,
): Promise<ArreteResolu> {
  const saisie = demande && dateArreteValide(demande) ? demande : null;
  const r = await appelerApi<{ arretes: string[] }>(`/arretes/${domaine}`, jeton ?? null);
  if (!r.ok) return { arrete: saisie, demande: null, disponibles: [], erreur: r.erreur };
  const arrete = resoudreArrete(saisie, r.donnees.arretes);
  return {
    arrete,
    demande: saisie && arrete && saisie !== arrete ? saisie : null,
    disponibles: r.donnees.arretes,
    erreur: null,
  };
}
